Reland "[HLSL] Implement the reflect HLSL function"

Icohedron opened the pull request:
This PR relands #122992.

A reland was attempted before (#123853), but it [failed to pass the `sanitizer-aarch64-linux-bootstrap-hwasan` buildbot](https://github.com/llvm/llvm-project/pull/123853#issuecomment-2608389396) due to the test `llvm/test/CodeGen/SPIRV/opencl/reflect-error.ll`

The issue has since been patched thanks to @vitalybuka, so the PR is safe to reland without any changes.
See https://github.com/llvm/llvm-project/pull/125599#discussion_r1966650839 and https://github.com/llvm/llvm-project/pull/125599#discussion_r1966650839

Icohedron committed:
Reapply "Reland "[HLSL] Implement the `reflect` HLSL function"" (#124046)

This reverts commit 0fe8e70c6609ff86cd40fbb45a85a8ed04c153c2.

Icohedron committed:
Mark hwasan as unsupported

Icohedron commented:
To get the PR up and running, I am just marking hwasan as unsupported for now.
A proper fix will be implemented later. The Issue #124045 will remain open to track this.

joaosaffran commented on clang/test/CodeGenHLSL/builtins/reflect.hlsl:
```diff
@@ -0,0 +1,177 @@
+// SPVCHECK-LABEL: define spir_func noundef nofpclass(nan inf) half @_Z17test_reflect_halfDhDh(
+// SPVCHECK-SAME: half noundef nofpclass(nan inf) [[I:%.*]], half noundef nofpclass(nan inf) [[N:%.*]]) local_unnamed_addr #[[ATTR0:[0-9]+]] {
+// SPVCHECK-NEXT:  [[ENTRY:.*:]]
+// SPVCHECK-NEXT:    [[MUL_I:%.*]] = fmul reassoc nnan ninf nsz arcp afn half [[I]], 0xH4000
```
Are those important to be tested? 

```suggestion
// SPVCHECK-NEXT:    [[MUL_I:%.*]] = fmul {{.*}} [[I]], 0xH4000
```

farzonl replied:
It looks like these are auto generated via `update_cc_test_checks.py` I woudn't take this change because it would just be reverted the next time someone runs `update_cc_test_checks.py`

joaosaffran commented on clang/test/CodeGenHLSL/builtins/reflect.hlsl:
```diff
@@ -0,0 +1,177 @@
+    return reflect(I, N);
+}
+
+// CHECK-LABEL: define noundef nofpclass(nan inf) <2 x half> @_Z18test_reflect_half2Dv2_DhS_(
```
```suggestion
// CHECK-LABEL: define noundef nofpclass(nan inf) <2 x half> @{{.*}}test_reflect_half2{{.*}}(
```

bogner commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -905,6 +903,14 @@ bool SPIRVInstructionSelector::selectExtInst(Register ResVReg,
                                              const SPIRVType *ResType,
                                              MachineInstr &I,
                                              GL::GLSLExtInst GLInst) const {
+  if (!STI.canUseExtInstSet(
+          SPIRV::InstructionSet::InstructionSet::GLSL_std_450)) {
+    std::string DiagMsg;
+    raw_string_ostream OS(DiagMsg);
+    I.print(OS, true, false, false, false);
+    DiagMsg += " is only supported with the GLSL extended instruction set.\n";
+    report_fatal_error(DiagMsg.c_str(), false);
+  }
```
IMO this isn't the right way to do this kind of error handling. `report_fatal_error` just causes the compiler to crash, but given that this can happen given IR for the wrong target just crashing because of it isn't great. We should probably be using `LLVMContext::diagnose` here instead so that we can bubble the error up to the driver.  Note that I do think we probably need to do a little bit of refactoring to use `diagnose` here, as this will probably just crash anyway with a `Cannot select` error if we just call diagnose and return from this function.

I'll comment on #123847 to this effect.

In the meantime, I also don't really like that we're adding this error handling only for the `GLSLExtInst` overload and not all of the extremely related cases. I think we should treat this as out of scope for the `reflect` function change, remove the `report_fatal_error`, and drop or XFAIL the `reflect-error.ll` test for now - we can reintroduce that test when we fix #123847.

farzonl replied:
I'm not sure I agree. The other intrinsic selections in  `selectIntrinsic` are doing `report_fatal_error` The only difference is I don't see the report fatal error code pathes are being tested. Switchinng to `LLVMContext::diagnose` would be a new pattern and if it requires a refactor like you claim we probably want to involve the SPIRV backend stake holders depending on how large of a refactor you imagine.

Second this is a case that should never be possible. The code that emits the reflect intrinsic is only emited when the target is `spirv` via the spirv target builtins. However the opencl test is  using the `spirv32` and `spirv64` targets. This intrinsic should be invalid if the target isn't `spirv` and to me it makes sense to fail immediately.

Further their is no mixed mode where you can do ExtInst in opencl and opengl so checking `canUseExtInstSet` seems like the right way forward.

Third I agree that the error handling in `GLSLExtInst` should have its reciprocal error in the `OpenCLExtInst`. However thats beyond the scope of this PR. I wouldn't want to add error handling like that without a test and I think a test like that is out of scope for a change about the `reflect` api

bogner replied:
> I'm not sure I agree. The other intrinsic selections in `selectIntrinsic` are doing `report_fatal_error` The only difference is I don't see the report fatal error code pathes are being tested. Switching to `LLVMContext::diagnose` would be a new pattern and if it requires a refactor like you claim we probably want to involve the SPIRV backend stake holders depending on how large of a refactor you imagine.

I am indeed asserting that many of the other uses of report_fatal_error in this file are also wrong. However, you're probably right that it makes sense to the existing pattern for now and clean this all up in a cohesive change.

> Second this is a case that should never be possible. The code that emits the reflect intrinsic is only emited when the target is `spirv` via the spirv target builtins. However the opencl test is using the `spirv32` and `spirv64` targets. This intrinsic should be invalid if the target isn't `spirv` and to me it makes sense to fail immediately.

If the case is impossible why isn't this just an assert? In any case "failing immediately" and "causing the program to crash" are different things, and `report_fatal_error` is the latter.

> Further their is no mixed mode where you can do ExtInst in opencl and opengl so checking `canUseExtInstSet` seems like the right way forward.
> 
> Third I agree that the error handling in `GLSLExtInst` should have its reciprocal error in the `OpenCLExtInst`. However thats beyond the scope of this PR. I wouldn't want to add error handling like that without a test and I think a test like that is out of scope for a change about the `reflect` api

IMO this is an argument for making a change that adds error handling to both functions *before* this PR, so that we don't end up in a state where the codebase is inconsistent.

Icohedron committed:
Re-support hwasan. Replace /dev/null with -

Since the error on hwasan was able to be reproduced locally and the "underlying" (though still not fully understood) issue was discovered, this commit re-adds support for running the test under hwasan and replaces the problematic llc output argument that was causing the test to fail under hwasan.

Icohedron commented:
Because the "underlying" -- though still not fully understood -- cause of the test failure under hwasan was found (refer to my [comment](https://github.com/llvm/llvm-project/issues/124045#issuecomment-2669906382) from issue #124045), I have re-added support for running the test under hwasan and fixed the issue with the RUN lines that made the test fail under hwasan in the first place.

I think this is the way forward, unless we think the error test should be omitted and relegated to a separate PR.

bogner reviewed (approved):
I have some misgivings about the way we're partially introducing the error handling here, but my disagreements are largely philosophical and procedural. Since we've worked around the issues with the hwasan bot at this point it's probably fine to get this PR in and leave the improvements to error handling to #124045

farzonl commented:
>I am indeed asserting that many of the other uses of report_fatal_error in this file are also wrong. However, you're probably right that it makes sense to the existing pattern for now and clean this all up in a cohesive change.

Thanks for clarifying. Making a comprehensive change works for me.

> If the case is impossible why isn't this just an assert? In any case "failing immediately" and "causing the program to crash" are different things, and report_fatal_error is the latter.

Its impossible from clang, but anyone can invoke this intrinsic from opt\llc and so wanted a fail fast that we could test so no one thought they could use this intrinsic if the extention mode was opencl.

> IMO this is an argument for making a change that adds error handling to both functions before this PR, so that we don't end up in a state where the codebase is inconsistent.

I actually agree with this, but this funnction was marked `[[maybe_unused]]` so felt we need a use before making alteration. Plus we need an opencl counter part to test and we didn't have that until  #124045 was filed.

vitalybuka commented on llvm/test/CodeGen/SPIRV/opencl/reflect-error.ll:
```diff
@@ -1,6 +1,5 @@
-; RUN: not llc -verify-machineinstrs -O0 -mtriple=spirv64-unknown-unknown %s -o /dev/null 2>&1 | FileCheck %s
-; RUN: not llc -verify-machineinstrs -O0 -mtriple=spirv32-unknown-unknown %s -o /dev/null 2>&1 | FileCheck %s
-; UNSUPPORTED: hwasan
+; RUN: not llc -verify-machineinstrs -O0 -mtriple=spirv64-unknown-unknown %s -o - 2>&1 | FileCheck %s
```
you don't need this patch
https://github.com/llvm/llvm-project/pull/128386/files passes: https://lab.llvm.org/buildbot/#/builders/55/builds/7459

vitalybuka reviewed (approved):
LGTM  regarding previous HWASAN crash, which is gone

Icohedron commented:
> LGTM regarding previous HWASAN crash, which is gone

Thank you for fixing the issue! I will revert the patches then.

Icohedron committed:
Revert "Re-support hwasan. Replace /dev/null with -"

This reverts commit b59d79d3f80a6814d325e48add5afaa78ac55560.

Icohedron committed:
Revert "Mark hwasan as unsupported"

This reverts commit ae25b96eb5481f7f96ffb88032e462cd435040ca.
